Plan Details: replace No Results view

ScoutHarris opened the pull request:
Fixes #9596 

This replaces the No Results view for Plan Details.

The No Results view is displayed in these cases:
- Loading plan detail
- Loading error
- No connection (this is a new addition)

To test:

This can be difficult to test because the plan details are loaded so fast. So I've included a hack for each scenario in case it's needed.

---

**Loading plan detail:**
Hack: 
- In `PlanDetailViewModel:tableViewModel` always return `ImmuTable.Empty`
- In `PlanDetailViewModel:noResultsViewModel` return the "Loading Plan..." model for the `ready` case.

Test:
- Go to site details > Plans > select a Plan.
- Verify the initial view is:
<img width="250" alt="loading_plan" src="https://user-images.githubusercontent.com/1816888/41492413-3ebbc682-70bc-11e8-9b55-b7edc7cd9ed3.png">

---

**Loading error:**
Hack: 
- In `PlanDetailViewModel:tableViewModel` always return `ImmuTable.Empty`
- In `PlanDetailViewModel:noResultsViewModel` return the "Oops" model for the `ready` case.

Test:
- Go to site details > Plans > select a Plan.
- Verify the view is:
<img width="250" alt="loading_error" src="https://user-images.githubusercontent.com/1816888/41492459-7f118d7a-70bc-11e8-8272-cda7f2c8c5b0.png">

---

**No connection:**
Hack: 
- In `PlanDetailViewModel:tableViewModel` always return `ImmuTable.Empty`
- In `PlanDetailViewModel:noResultsViewModel` return the "No connection" model for the `ready` case.

Test:
- Go to site details > Plans.
- Disable internet connection (unless you've implemented the hack).
- Select a Plan.
- Verify the view is:
<img width="250" alt="no_connection" src="https://user-images.githubusercontent.com/1816888/41492555-0eab62da-70bd-11e8-9705-632258475522.png">

etoledom commented on WordPress/Classes/ViewRelated/Plans/PlanDetailViewController.swift:
```diff
@@ -247,6 +226,65 @@ class PlanDetailViewController: UIViewController {
+        noResultsViewController.bindViewModel(viewModel)
+
+        noResultsViewController.view.frame = noResultsViewFrame()
+        noResultsViewController.view.backgroundColor = UIColor.white
```
Small detail: `UIColor` can 💥 for less verbosity

ScoutHarris replied:
Ya know, I know that. And one day I'll stop doing it. 😉

etoledom commented on WordPress/Classes/ViewRelated/Plans/PlanDetailViewModel.swift:
```diff
@@ -41,20 +41,22 @@ struct PlanDetailViewModel {
+            } else {
+                return NoResultsViewController.Model(title: NSLocalizedString("No connection", comment: ""),
+                                                     subtitle: NSLocalizedString("An active internet connection is required to view plans", comment: ""))
+            }
```
Just for cleanness: could we here create a helper method that returns the `errorModel`? 
Something like `return errorViewModel()`

ScoutHarris replied:
The helper method was going to be almost as bad, so I moved the strings to a struct to make it more readable.

etoledom commented on WordPress/Classes/ViewRelated/Plans/PlanDetailViewController.swift:
```diff
@@ -247,6 +226,65 @@ class PlanDetailViewController: UIViewController {
+        tableView.addSubview(withFadeAnimation: noResultsViewController.view)
+        addChildViewController(noResultsViewController)
+
+        noResultsViewController.didMove(toParentViewController: self)
```
This process of adding a child view controller is also something we would have to do always. Maybe we can hide it too in the `NoResultsViewController` class?
I'm not sure if the implementations will always be the same for every case but it would be cool if that's possible :)

ScoutHarris replied:
Hey @etoledom . You noticed the repetitiveness did ya? 😄 Yea, I did too. And that was actually going to be my next task - extract out as much of that as possible to the `NoResultsViewController` class. I just didn't make it part of this PR because I wanted to look at it further. Just as you said - not implementations will be the same.  So I wanted to flush out the patterns before doing so.

etoledom replied:
Perfect! ✅

ScoutHarris commented on WordPress/Classes/ViewRelated/Plans/PlanDetailViewController.swift:
```diff
@@ -247,6 +226,65 @@ class PlanDetailViewController: UIViewController {
+    private func setupNoResultsViewController() {
+        let noResultsStoryboard = UIStoryboard(name: "NoResults", bundle: nil)
+        guard let noResultsViewController = noResultsStoryboard.instantiateViewController(withIdentifier: "NoResults") as? NoResultsViewController else {
+            return
```
Refer to my comment re repetitiveness.

ScoutHarris commented:
OK, @etoledom I think it's ready for you again.

The primary change is the NoResultsView is now added to the table as a cell instead of a subview.

There are still a couple of issues that may affect the layout, which I will address separately:
- Sometimes the labels are cut off
  - This is noted in https://github.com/wordpress-mobile/WordPress-iOS/issues/9606.
- Sometimes the top of the image is cut off.
  - This is related to the image having a lot of white space, making it difficult to layout properly. I've asked Design to provide a new cropped image.

Please note I've moved this PR to the next release (10.5), so I'll have time to fix those before this goes live.

Thank you!!
